source/common/upstream/upstream_impl.cc

```diff
@@ -1608,13 +1611,24 @@ bool BaseDynamicClusterImpl::updateDynamicHostList(
     // (currently there is only one criterion, but we might add more in the future):
     // - The cluster health checker is activated and a new host is matched with the existing one,
     //   but the health check address is different.
-    const bool skip_inplace_host_update =
-        health_checker_ != nullptr && existing_host_found &&
-        *existing_host->second->healthCheckAddress() != *host->healthCheckAddress();
+    const bool health_check_address_changed =
+        (health_checker_ != nullptr && existing_host_found &&
+         *existing_host->second->healthCheckAddress() != *host->healthCheckAddress());
+
+    const bool locality_changed =
+        (existing_host_found &&
+         ((host->locality().region() != existing_host->second->locality().region()) ||
+          (host->locality().zone() != existing_host->second->locality().zone()) ||
+          (host->locality().sub_zone() != existing_host->second->locality().sub_zone())));
+    if (locality_changed) {
+      existing_hosts_for_current_priority_locality_changed.emplace(existing_host->first);
+    }
```

Comment:
You can probably just write
`if (existing_host_found && host->locality() != existing_host->second->locality()) {
  existing_hosts_for_current_priority_locality_changed.emplace(existing_host->first);
 }`